Assets/SEE/Net/Network.cs

```diff
@@ -629,153 +573,201 @@ private void OnSceneUnloaded(Scene scene)
+        {
+            if (RoomPassword == Encoding.ASCII.GetString(request.Payload))
             {
-                float waitingTime = 0;
-                float waitingTimePerIteration = 0.5f;
+                Debug.Log($"Client {request.ClientNetworkId} has sent right room password.\n");
+                response.Approved = true;
 
-                while (!NetworkManager.Singleton.IsConnectedClient)
-                {
-                    Debug.Log($"Client is waiting for connection to server {ServerAddress} {waitingTime}/{maxWaitingTime}...\n");
-                    yield return new WaitForSeconds(waitingTimePerIteration);
-                    waitingTime += waitingTimePerIteration;
-                    if (waitingTime > maxWaitingTime)
-                    {
-                        break;
-                    }
-                }
-                if (NetworkManager.Singleton.IsConnectedClient)
-                {
-                    callBack(true, $"Client is connected to server {ServerAddress}.");
-                }
-                else
-                {
-                    callBack(false, $"Could not connect to server {ServerAddress}.");
-                    throw new NoServerConnection($"Could not connect to server {ServerAddress}");
-                }
+            }
+            else
+            {
+                response.Approved = false;
+                response.Reason = "Invalid password";
+                Debug.LogWarning($"Client {request.ClientNetworkId} has sent wrong room password.\n");
             }
         }
 
         /// <summary>
-        /// The maximal waiting time in seconds a client is willing to wait until a connection
-        /// can be established.
+        /// Removes the reference to the callback used to send the client back to the main menu
+        /// because the connection was successfully established.
+        /// The <paramref name="owner"/> is not used.
         /// </summary>
-        private const float maxWaitingTime = 5 * 60;
+        /// <param name="owner">ID of the owner (ignored)</param>
+        private void OnClientConnectedCallback(ulong owner)
+        {
+            callbackToMenu?.Invoke(true, $"You are connected to {ServerAddress}.\n");
+            callbackToMenu = null;
+        }
+
+        /// <summary>
+        /// Sends the client back to the main menu because the connection could not
+        /// be established.
+        /// The <paramref name="owner"/> is not used.
+        /// </summary>
+        /// <param name="owner">ID of the owner (ignored)</param>
+        private void OnClientDisconnectCallback(ulong owner)
+        {
+            callbackToMenu?.Invoke(false,
+                                   $"The server {ServerAddress} has refused the connection due to the following reason: "
+                                     + NetworkManager.Singleton.DisconnectReason);
```

Comment:
Is this thread-safe, or is it possible that the `DisconnectReason` has changed in the meantime?